nomad/eval_broker.go

```diff
@@ -586,15 +592,28 @@ func (b *EvalBroker) Ack(evalID, token string) error {
 
 	// Check if there are any blocked evaluations
 	if blocked := b.blocked[namespacedID]; len(blocked) != 0 {
-		raw := heap.Pop(&blocked)
+
+		// Any blocked evaluations with ModifyIndexes older than the just-ack'd
+		// evaluation are no longer useful, so it's safe to drop them.
+		cancelable := blocked.MarkForCancel(recentlyAckedIndex)
+		b.cancelable = append(b.cancelable, cancelable...)
+		b.stats.TotalCancelable = b.cancelable.Len()
+		b.stats.TotalBlocked -= cancelable.Len()
+
+		// If any remain, enqueue an eval
+		if len(blocked) > 0 {
+			raw := heap.Pop(&blocked)
+			eval := raw.(*structs.Evaluation)
```

Comment:
Surely @shoenig has a generic heap for us to avoid having to assert a type? :grin: